### ISSUE TITLE
GitHub Desktop “Can't Find” Local Repository That Exists

### ISSUE THREAD

**Issue by patrickdark** [desktop/desktop#2851]
### Description

GitHub Desktop says “Can't find "gecko-dev"” “It was last seen at C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev\” with “Locate…”, “Clone”, and “Remove” buttons after syncing a repository with the online version.

Clicking the “Locate…” button opens the “gecko-dev” directory containing all of the relevant files…

### Version

I was running 1.0.0. An automatic update and restart to 1.0.1 brought me to the same message.

**GitHub Desktop version:** See the preceding “Version” information.

**OS version:** Microsoft Windows [Version 10.0.15063]

### Steps to Reproduce

1. Activate the “File” menu: “Clone repository…” command.
2. Click an online repository from the list.
3. Set the “Local path” field to the pre-existing local folder for that repository.
4. Click the “Clone” button.
5. The error message from https://github.com/desktop/desktop/issues/2838 appears.
6. Click the “Close” button.
7. Click the “Select a repository” button.
8. Select the name of the same repository from the list.
9. The aforementioned error message appears.

The repository is my clone of the “gecko-dev” (Mozilla Firefox) repository, which takes hours to sync since it’s about 5 GB of files, so I haven’t tried syncing again yet.

**Expected behavior:** I’m not sure what should happen.

**Actual behavior:** See item item 9.

**Reproduces how often:** 100%

### Logs

The log file is about 2 MB in size. This is the beginning and end of it:

2017-09-22T05:14:45.741Z - warn: [ui] Unknown URL action: unknown - payload: {"name":"unknown","url":"--squirrel-firstrun"}
2017-09-22T05:14:47.719Z - info: [ui] Executing installGlobalLFSFilter: git lfs install --skip-repo (took 1.954s)
2017-09-22T05:19:13.263Z - error: [ui] `git -c credential.helper= lfs clone --recursive --progress --progress -- https://github.com/patrickdark/gecko-dev.git C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev` exited with an unexpected code: 2.
fatal: destination path 'C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev' already exists and is not an empty directory.
Error(s) during clone:
git clone failed: exit status 128

2017-09-22T05:20:35.771Z - info: [ui] Opt in reported.
2017-09-22T05:20:36.078Z - info: [ui] Stats reported.
2017-09-22T05:21:40.288Z - error: [ui] `git -c credential.helper= lfs clone --recursive --progress --progress -- https://github.com/patrickdark/gecko-dev.git C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev` exited with an unexpected code: 2.
fatal: destination path 'C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev' already exists and is not an empty directory.
Error(s) during clone:
git clone failed: exit status 128

2017-09-22T05:22:05.026Z - info: [ui] Executing isUsingLFS: git lfs track (took 5.927s)
2017-09-22T14:25:24.021Z - info: [ui] Executing fetch: git -c credential.helper= fetch --progress --prune origin (took 32591.439s)
2017-09-22T14:25:24.128Z - error: [ui] `git -c credential.helper= fetch --progress --prune origin` exited with an unexpected code: 128.
remote: Counting objects: 6612900           
remote: Counting objects: 6612908, done.        
remote: Compressing objects:  14% (1/7)           
remote: Compressing objects:  28% (2/7)           
remote: Compressing objects:  42% (3/7)           
remote: Compressing objects:  57% (4/7)           
remote: Compressing objects:  71% (5/7)           
remote: Compressing objects:  85% (6/7)           
remote: Compressing objects: 100% (7/7)           
remote: Compressing objects: 100% (7/7), done.        
Receiving objects:   0% (1/6612908) 
  
[…]

Receiving objects:  52% (3468667/6612908), 2.53 GiB | 90.00 KiB/s   
Receiving objects:  52% (3469049/6612908), 2.53 GiB | 106.00 KiB/s   
Receiving objects:  52% (3469236/6612908), 2.53 GiB | 126.00 KiB/s   
Receiving objects:  52% (3469370/6612908), 2.53 GiB | 133.00 KiB/s   
Receiving objects:  52% (3469573/6612908), 2.53 GiB | 146.00 KiB/s   
Receiving objects:  52% (3469812/6612908), 2.53 GiB | 144.00 KiB/s   
Receiving objects:  52% (3470054/6612908), 2.53 GiB | 143.00 KiB/s   
Receiving objects:  52% (3470151/6612908), 2.53 GiB | 141.00 KiB/s   
Receiving objects:  52% (3470250/6612908), 2.53 GiB | 135.00 KiB/s   
Receiving objects:  52% (3470998/6612908), 2.53 GiB | 134.00 KiB/s   
Receiving objects:  52% (3471372/6612908), 2.53 GiB | 121.00 KiB/s   
Receiving objects:  52% (3472115/6612908), 2.53 GiB | 117.00 KiB/s   
Receiving objects:  52% (3472403/6612908), 2.53 GiB | 113.00 KiB/s   
Receiving objects:  52% (3472535/6612908), 2.53 GiB | 100.00 KiB/s   
Receiving objects:  52% (3472958/6612908), 2.53 GiB | 107.00 KiB/s   
Receiving objects:  52% (3473216/6612908), 2.53 GiB | 89.00 KiB/s   
Receiving objects:  52% (3473431/6612908), 2.53 GiB | 84.00 KiB/s   
Receiving objects:  52% (3473690/6612908), 2.53 GiB | 75.00 KiB/s   
fatal: early EOF
fatal: The remote end hung up unexpectedly
fatal: index-pack failed
error: RPC failed; curl 56 Send failure: Connection was reset

(The error was parsed as 4: The remote disconnected. Check your Internet connection and try again.)

2017-09-22T14:25:24.403Z - error: [ui] `git status --untracked-files=all --branch --porcelain=2 -z` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T14:25:24.422Z - error: [ui] `git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T22:11:10.904Z - error: [ui] Error checking for updates
Error: GitHub Desktop was not able to contact the update server. Ensure you have internet connectivity and try again.
    at Object.t.parseError (file:///C:/Users/Patrick%20Dark/AppData/Local/GitHubDesktop/app-1.0.0/resources/app/renderer.js:14:297491)
    at onAutoUpdaterError (file:///C:/Users/Patrick%20Dark/AppData/Local/GitHubDesktop/app-1.0.0/resources/app/renderer.js:5:303332)
    at CallbacksRegistry.apply (C:\Users\Patrick Dark\AppData\Local\GitHubDesktop\app-1.0.0\resources\electron.asar\common\api\callbacks-registry.js:48:25)
    at EventEmitter.<anonymous> (C:\Users\Patrick Dark\AppData\Local\GitHubDesktop\app-1.0.0\resources\electron.asar\renderer\api\remote.js:299:21)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:197:7)
2017-09-22T22:12:05.745Z - error: [ui] `git status --untracked-files=all --branch --porcelain=2 -z` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T22:12:05.798Z - error: [ui] `git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T22:12:30.312Z - info: [ui] Executing installGlobalLFSFilter: git lfs install --skip-repo (took 1.239s)
2017-09-22T22:59:28.410Z - info: [ui] launching: 1.0.1 (Windows 10.0.15063)
2017-09-22T22:59:29.498Z - info: [ui] Executing installGlobalLFSFilter: git lfs install --skip-repo (took 1.215s)
2017-09-22T22:59:40.171Z - error: [ui] `git status --untracked-files=all --branch --porcelain=2 -z` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T22:59:40.174Z - error: [ui] `git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T22:59:40.384Z - error: [ui] `git remote -v` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T23:12:53.190Z - error: [ui] `git status --untracked-files=all --branch --porcelain=2 -z` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T23:12:53.204Z - error: [ui] `git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T23:12:53.421Z - error: [ui] `git remote -v` exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
2017-09-22T23:13:52.269Z - error: [ui] `git -c credential.helper= lfs clone --recursive --progress --skip-repo --progress -- https://github.com/patrickdark/gecko-dev.git C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev` exited with an unexpected code: 2.
fatal: destination path 'C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev' already exists and is not an empty directory.
Error(s) during clone:
git clone failed: exit status 128

2017-09-22T23:16:13.251Z - error: [ui] `git -c credential.helper= lfs clone --recursive --progress --skip-repo --progress -- https://github.com/patrickdark/gecko-dev.git C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev` exited with an unexpected code: 2.
fatal: destination path 'C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev' already exists and is not an empty directory.
Error(s) during clone:
git clone failed: exit status 128

2017-09-22T23:18:59.604Z - error: [ui] `git -c credential.helper= lfs clone --recursive --progress --skip-repo --progress -- https://github.com/patrickdark/gecko-dev.git C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev` exited with an unexpected code: 2.
fatal: destination path 'C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev' already exists and is not an empty directory.
Error(s) during clone:
git clone failed: exit status 128

#### Additional Information

This was on a Windows 10 local account without administrator privileges. The “gecko-dev” repository was copied from another hard drive.

**Comment by shiftkey** [comment]
@patrickdark thanks for the log file. Let me walk through it now to understand what's happening:

> 4. Click the “Clone” button.
> 5. The error message from #2838 appears.

The error from the log for the first clone looks like we don't warn you about an existing repository at that path - that's #2830:

```
fatal: destination path 'C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev' already exists and is not an empty directory.
```

Then it looks like it tries to fetch from some repository, but fails (due to some size limit?):

```
2017-09-22T14:25:24.021Z - info: [ui] Executing fetch: git -c credential.helper= fetch --progress --prune origin (took 32591.439s)
2017-09-22T14:25:24.128Z - error: [ui] git -c credential.helper= fetch --progress --prune origin exited with an unexpected code: 128.
...
Receiving objects: 52% (3473690/6612908), 2.53 GiB | 75.00 KiB/s
fatal: early EOF
fatal: The remote end hung up unexpectedly
fatal: index-pack failed
error: RPC failed; curl 56 Send failure: Connection was reset
```

Then it looks like we're stuck in the "not a git repository" state. I suspect the folder exists but something is missing from the `.git` folder which is causing Git to error in this way:

```
2017-09-22T14:25:24.403Z - error: [ui] git status --untracked-files=all --branch --porcelain=2 -z exited with an unexpected code: 128.
fatal: Not a git repository (or any of the parent directories): .git

(The error was parsed as 27: This is not a git repository.)
```

Let's take a step back to that original clone error:

> The “gecko-dev” repository was copied from another hard drive.

This is probably the reason the initial error when cloning the repository to a folder that already exists at a given location.

Let's see if we can get you to a happy place. Could you try this:

 - remove the "missing" `gecko-dev` from Desktop
 - cleanup whatever is under the local `gecko-dev` folder on disk
 - bring over a new copy of the `gecko-dev` folder from the other hard drive
 - open **File** | **Add existing repository** in Desktop
 - add this `gecko-dev` folder

I think that and fixing #2830 should prevent this from happening again.

Let me know if that works for you!

**Comment by patrickdark** [comment]
@shiftkey I’m guessing you mean “File” menu: “Clone repository” command by “File | Add existing repository” since the latter command doesn’t exist. I attempted steps similar to the ones you requested:

1. Remove the “/GitHub Repositories/gecko-dev/” directory pair to another location.
2. Remove “gecko-dev” from the repository list by right-clicking it and selecting the “Remove” command.
3. Using the “File” menu: “Clone repository” command of the same repository to the same location. This created a hidden “.git” folder in the empty “gecko-dev” folder. (This folder was missing from my backed-up “gecko-dev” folder probably because my Western Digital backup software purposely excludes files with the operating system’s hidden attribute.)
4. Since there’s no way to pause the syncing process, I force-exited GitHub Desktop by pressing the “Close” button during the syncing process.
5. I moved all of the files from the removed “gecko-dev” folder into the new “gecko-dev” folder (with the new hidden “.git” folder).
6. I restarted GitHub Desktop.
7. I had to repeat step 3 since the repository was missing from the repository list.
8. Syncing completed*, then I ended up with the same error message that caused me to file this issue: “Cannot find "gecko-dev"”. (*I didn’t actually witness syncing complete, but I noted that it got to at least 95% and there was no error log indicating a download failure this time.)

During the syncing process, I was repeatedly bombarded with error message dialogs stating the following: “The output from the command could not fit into the allocated `stdout` buffer. Set `options.maxBuffer` to a larger value than 10485760 bytes.”

Clicking the “Close” button on that error message dialog would just result in a new such message being displayed immediately or a short time later. Clicking the “Close” button or leaving the messages alone seemed to have no effect on the syncing progress meter.

I experienced the aforementioned error messages on the first sync attempt as well but didn’t mention them since I didn’t have a screenshot of them until now.

**Comment by shiftkey** [comment]
> I’m guessing you mean “File” menu: “Clone repository” command by “File | Add existing repository” since the latter command doesn’t exist. 

Apologies, I meant **File** | **Add local repository**:

<img width="346" src="https://user-images.githubusercontent.com/359239/30844575-fdc8ae86-a2d2-11e7-817a-8d3cdf115e71.png">

Then click **Choose** and find your `gecko-dev` folder on disk to add it to the app:

<img width="993" alt="screen shot 2017-09-26 at 3 57 12 pm" src="https://user-images.githubusercontent.com/359239/30844670-63c36b86-a2d3-11e7-9ef3-7e413b746722.png">

This bypasses all the potential issues about cloning or fetching from a remote, which we can dig into once we're tracking the repository correctly in the app.

> During the syncing process, I was repeatedly bombarded with error message dialogs stating the following: “The output from the command could not fit into the allocated `stdout` buffer. Set `options.maxBuffer` to a larger value than 10485760 bytes.”

We're tracking this one in #2449 but I don't have a good answer for you on this right now.

**Comment by patrickdark** [comment]
@shiftkey That doesn’t seem to work.

If I follow the above steps after removing the remote repository, I’m repeatedly shown #2449 error messages even though GitHub Desktop doesn’t appear to be doing anything.

Furthermore, changes aren’t recorded, which sort of defeats the point. (Changes _are_ recorded for a tiny repository for a Firefox extensiont that syncs as expected with its online repository, so I know what it should look like.)

**Comment by shiftkey** [comment]
@patrickdark is it this repository you're trying to work with? https://github.com/mozilla/gecko-dev

**Comment by patrickdark** [comment]
@shiftkey Yes, well, my clone of it, which only has one commit.

**Comment by shiftkey** [comment]
@patrickdark I can have a go at working with that repository to see if I can fix #2449

**Comment by patrickdark** [comment]
@shiftkey Good luck! At 5 GB and 6.6 million files, it’ll probably take hours just to download it.

**Comment by shiftkey** [comment]
@patrickdark so the vanilla `gecko-dev` repository doesn't trigger the issue in the same way for me. This detail is interesting:

> Yes, well, my clone of it, which only has one commit.

Could you describe the change you've made here compared to the original repository? Squashed all the commits together?

**Comment by patrickdark** [comment]
@shiftkey I deleted and moved some lines in one CSS file (in layout/style/res/html.css). I wouldn’t expect that to result in any difference.

**Comment by patrickdark** [comment]
@shiftkey I wonder if this is related to running GitHub desktop on a non-administrator account. I launched GitHub Desktop as an administrator (i.e., on the same account but with the password of the system administrator account) and started another clone of “gecko-dev” and (A) wasn’t prompted about a directory already existing and (B) haven’t received any error messages yet at 10% into the cloning process.

I noticed that I had to log-in again and that all repository lists were missing on log-in despite logging in from the same account, which seems to potentially be another bug.

Or maybe a fresh log-in eliminated problems from 1.0.0. I don’t know.

**Comment by shiftkey** [comment]
@patrickdark I think we're a bit all over the place now. Let me call out the specific issues I'm still interested in.

```
fatal: destination path 'C:\Users\Patrick Dark\Documents\GitHub Repositories\gecko-dev' already exists and is not an empty directory.
Error(s) during clone:
git clone failed: exit status 128
```

I believe we've addressed this in #2863 which should actively warn you about the folder existing before cloning. **I'm not going to worry about this one.**

```
Receiving objects: 52% (3473690/6612908), 2.53 GiB | 75.00 KiB/s
fatal: early EOF
fatal: The remote end hung up unexpectedly
fatal: index-pack failed
error: RPC failed; curl 56 Send failure: Connection was reset
```

This still feels like an intermittent network issue (around the same time our update check fails).

```
2017-09-22T22:11:10.904Z - error: [ui] Error checking for updates
Error: GitHub Desktop was not able to contact the update server. Ensure you have internet connectivity and try again.
```

**I'm not going to focus on this one either.**

> During the syncing process, I was repeatedly bombarded with error message dialogs stating the following: “The output from the command could not fit into the allocated stdout buffer. Set options.maxBuffer to a larger value than 10485760 bytes.”

This is a generic message from Node and is likely due to Git emitting too much output to fit in it's buffer. We have a workaround for this for diffs by using the spawn API and streaming the output, but we don't use it elsewhere currently. It's also not clear to me the context for this message as "sync" doesn't map to a Git action. Can you confirm what you were doing to trigger this action - cloning a repository, adding an existing folder, or something else? **I am interested in this issue and making this error go away.**

> Furthermore, changes aren’t recorded. (Changes are recorded for a tiny repository for a Firefox extension that syncs as expected with its online repository, so I know what it should look like.)

I'm not clear on what you mean here by "recorded". **Please open a fresh issue with details about recreating this.**

> I wonder if this is related to running GitHub Desktop on a non-administrator account. I launched GitHub Desktop as an administrator (i.e., on the same account but with the password of the system administrator account) and started another clone of “gecko-dev” and (A) wasn’t prompted about a directory already existing and (B) haven’t received any error messages yet at 10% into the cloning process.

I don't believe it's related to the account you're running in, but I'm glad to see it seems to be working. It's not clear to me which version you're running now (1.0.2 has the fix I mentioned above).

> I noticed that I had to log in to GitHub again and that all repository lists were missing on log-in despite logging in from the same Windows account, which seems to potentially be another bug.

We use the Windows Credential store to track your credentials, so elevating the running process might affect what credentials you can access. **I'm not interested in this right now, but if you care about this please file a separate issue with details to recreate it.**

**Comment by joshaber** [comment]
Thank you for your issue!

We haven't gotten a response to our questions above. With only the information that is currently in the issue, we don't have enough information to take action. I'm going to close this but don't hesitate to reach out if you have or find the answers we need, we'll be happy to reopen the issue.
